cpp/utils.cpp

```diff
@@ -1,8 +1,248 @@
-// Copyright (C) 2021 Jørgen S. Dokken
+// Copyright (C) 2021 Jørgen S. Dokken and Sarah Roggendorf
 //
 // This file is part of DOLFINx_CONTACT
 //
 // SPDX-License-Identifier:    MIT
 #include "utils.h"
 
 using namespace dolfinx_contact;
+
+//-----------------------------------------------------------------------------
+void dolfinx_contact::pull_back(xt::xtensor<double, 3>& J,
+                                xt::xtensor<double, 3>& K,
+                                xt::xtensor<double, 1>& detJ,
+                                const xt::xtensor<double, 2>& x,
+                                xt::xtensor<double, 2>& X,
+                                const xt::xtensor<double, 2>& coordinate_dofs,
+                                const dolfinx::fem::CoordinateElement& cmap)
+{
+  // number of points
+  const std::size_t num_points = x.shape(0);
+  assert(J.shape(0) >= num_points);
+  assert(K.shape(0) >= num_points);
+  assert(detJ.shape(0) >= num_points);
+
+  // Get mesh data from input
+  const size_t tdim = K.shape(1);
+
+  // -- Lambda function for affine pull-backs
+  auto pull_back_affine
```

Comment:
This lambda has to be refactored, ref.
https://github.com/Wells-Group/asimov-contact/issues/37
and in particular:
https://github.com/FEniCS/dolfinx/pull/1937/files#diff-650e00bb71aa07a04c1edde359b3552e183f3b16652afb7cf45f1ed672c22642R312-R318